src/TorchSharp/DisposeScope.cs

```diff
@@ -0,0 +1,296 @@
+            Detach(new IDisposable[] { disposable });
+            return disposable;
+        }
+
+        /// <summary>
+        /// Detaches/excludes a set of tensors/disposables from the all dispose scopes, see overloaded methods. See MoveToOuter
+        /// if you wish to move it to the outer dispose scope.
+        /// </summary>
+        public (T1 first, T2 second) Detach<T1, T2>(T1 first, T2 second)
+            where T1 : IDisposable where T2 : IDisposable
+        {
+            Detach(new IDisposable[] { first, second });
+            return (first, second);
+        }
+
+        /// <summary>
+        /// Detaches/excludes a set of tensors/disposables from the all dispose scopes, see overloaded methods. See MoveToOuter
+        /// if you wish to move it to the outer dispose scope.
+        /// </summary>
+        public (T1 first, T2 second, T3 third) Detach<T1, T2, T3>(T1 first, T2 second, T3 third)
+            where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+        {
+            Detach(new IDisposable[] { first, second, third });
+            return (first, second, third);
+        }
+
+        /// <summary>
+        /// Detaches/excludes a set of tensors/disposables from the all dispose scopes, see overloaded methods. See MoveToOuter
+        /// if you wish to move it to the outer dispose scope.
+        /// </summary>
+        public void Detach(params IDisposable[] disposables) => Detach((IEnumerable<IDisposable>)disposables);
+
+        /// <summary>
+        /// Detaches/excludes a set of tensors/disposables from the all dispose scopes, see overloaded methods. See MoveToOuter
+        /// if you wish to move it to the outer dispose scope.
+        /// </summary>
+        public void Detach(IEnumerable<IDisposable> disposables)
+        {
+            foreach (var disposable in disposables) {
+                if (Disposables.Remove(disposable)) {
+                    _disposeScopeManager.StatisticsInstance.DetachedFromScopeCount++;
+                    if (disposable is torch.Tensor tensor) {
+                        tensor.OwningDisposeScope = null;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Disposes everything currenly in the dispose scope.
+        /// </summary>
+        public void DisposeEverything() => DisposeEverythingBut(Enumerable.Empty<IDisposable>());
+
+        /// <summary>
+        /// As an intermediate step, you can dispose all the tensors/disposables currently scheduled for dispose, to
+        /// clear up some memory without creating a new scope. Note that this doesn't permanently exclude the
+        /// tensors from disposing, use Exclude for that. Also, excluded tensors don't need to be included
+        /// here.
+        /// </summary>
+        public void DisposeEverythingBut(IEnumerable<IDisposable> inKeep)
+        {
+            // Avoiding multiple enumerations
+            var keep = inKeep.ToList();
```

Comment:
When does the disposable get removed from the list? Does it?

Reply:
Clearly a bug - and I had no test to detect it. Fixed the bug, added a test.